clippy_lints/src/functions/hidden_static_lifetime.rs

```diff
@@ -0,0 +1,280 @@
+								&format!("try removing the lifetime parameter `{}` and changing references to `'static`", generic.name.ident().as_str()));
+									};
+								};
+							};
+						};
+
+						// Check again.
+						if !lifetime_is_used &&
+							check_validness(ret_ty, generic, generics) &&
+							lifetime_not_used_in_ty(ret_ty, generic) {
+								span_lint_and_help(cx,
+								HIDDEN_STATIC_LIFETIME,
+								generic.span,
+								"this lifetime can be changed to `'static`",
+								None,
+								&format!(
+									"try removing the lifetime parameter `{}` and changing references to `'static`",
+									generic.name.ident().as_str()),
+							);
+						};
+					};
+				};
+			};
+        };
+    };
+}
+
+fn check_if_uses_lifetime(input: &Ty<'_>, generic_name: &ParamName) -> bool {
+    if let TyKind::Path(QPath::Resolved(_, path)) = &input.kind {
+        for segment in path.segments {
+            for arg in segment.args().args {
+                // If input's lifetime and function's are the same.
+                if let GenericArg::Lifetime(lifetime) = arg {
+                    if lifetime.is_anonymous() {
+                        return true;
+                    }
+                    if let ParamName::Plain(ident) = generic_name {
+                        if lifetime.ident.name == ident.name {
+                            return true;
+                        };
+                    };
+                };
+            }
+        }
+    };
+    false
+}
+
+fn check_validness(ret_ty: &Ty<'_>, generic: &GenericParam<'_>, generics: &Generics<'_>) -> bool {
+    // (#10123) Comment by @Jarcho explains what "valid" means.
+    // The lint doesn't check invalid return types, because not every lifetime can be changed to 'static
+    // without problems.
+    if let TyKind::BareFn(barefn) = ret_ty.peel_refs().kind {
+        for input in barefn.decl.inputs {
+            if let TyKind::Ref(lifetime, _) = input.kind {
+                if lifetime.ident.name == generic.name.ident().name {
+                    return false;
+                }
+            }
+        }
+    }
+
+    if let TyKind::Ref(_, mut_ty) = &ret_ty.kind {
+        // Check for &'a mut &'b T
+        if mut_ty.mutbl.is_mut() {
+            // This path diverges:
+            // * &'a mut &'b T : Not valid
+            // * &'a mut T : Only if T: 'static (Checked before)
+            if let TyKind::Ref(_, _) = &mut_ty.ty.kind {
+                return false;
+            };
+
+            for predicate in generics.predicates {
+                for bound in predicate.bounds() {
+                    if let GenericBound::Outlives(lifetime) = bound {
+                        if !lifetime.is_static() {
+                            return false;
+                        }
+                    }
+                }
+            }
+        }
+    };
+    true
+}
+
+// true = lifetime isn't used (success)
+fn lifetime_not_used_in_ty(ret_ty: &Ty<'_>, generic: &GenericParam<'_>) -> bool {
```

Comment:
This should be merged with the previous function. While the type is a reference you're unwrapping it looking for lifetimes to change to `'static`. Once you hit a type which isn't a reference you're then looking to see if the lifetime is used again.